Objects/moduleobject.c

```diff
@@ -333,6 +348,27 @@ PyModule_FromDefAndSpec2(PyModuleDef* def, PyObject *spec, int module_api_versio
         goto error;
     }
 
+#ifdef Py_GIL_DISABLED
+    PyThreadState *tstate = _PyThreadState_GET();
+    const PyConfig *config = _PyInterpreterState_GetConfig(interp);
+    if (gil_slot == Py_MOD_GIL_USED && config->enable_gil == _PyConfig_GIL_DEFAULT) {
```

Comment:
As noted above, there's a strong relationship between `Py_mod_multiple_interpreters` and `Py_mod_gil`.  I think the code should reflect that here:

```suggestion
    if (has_gil_slot) {
        if (has_multiple_interpreters_slot) {
            if (gil_slot == Py_MOD_GIL_USED && multiple_interpreters == Py_MOD_PER_INTERPRETER_GIL_SUPPORTED) {
                // Warn about this mismatch?
            }
            else if (gil_slot == Py_MOD_GIL_NOT_USED && multiple_interpreters == Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED) {
                // Warn about this mismatch?
            }
        }
        else {
            // Imply?
            multiple_interpreters = gil_slot == Py_MOD_GIL_USED
                ? Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED
                : Py_MOD_PER_INTERPRETER_GIL_SUPPORTED;
    }
#ifdef Py_GIL_DISABLED
    PyThreadState *tstate = _PyThreadState_GET();
    const PyConfig *config = _PyInterpreterState_GetConfig(interp);
    if (gil_slot == Py_MOD_GIL_USED && config->enable_gil == _PyConfig_GIL_DEFAULT) {
```

Reply:
I don't think either of those are actually mismatches:
- Line 354: This means that multiple interpreters (with their own GILs) can be running in this module concurrently, but only one thread per interpreter at a time. This could be as straightforward as each interpreter's copy of the module state containing a data structure that isn't thread-safe. In a free-threaded build, we'll already warn the first time any interpreter loads a module like this.
- Line 357: This is a little weirder, but it doesn't seem completely unreasonable to have thread-safe global state that can't be partitioned by interpreter (due to design restrictions of a third-party library, for example).

I'm confused by line 363 (and 351, by extension). My reading of your suggested code is "if the module sets `Py_mod_gil` to `Py_MOD_GIL_USED` and doesn't provide `Py_mod_multiple_interpreters`, default `Py_mod_multiple_interpreters` to `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`". In your [comment on `module.rst`](https://github.com/python/cpython/pull/116882/files#r1538106137), you said that the default is `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`. That lines up with what [the code does](https://github.com/python/cpython/blob/0821923aa979a72464c5da8dfa53a719bba5801c/Objects/moduleobject.c#L316-L320), but [the docs say](https://github.com/python/cpython/blob/0821923aa979a72464c5da8dfa53a719bba5801c/Doc/c-api/module.rst?plain=1#L409-L410) that the default is `Py_MOD_MULTIPLE_INTERPRETERS_NOT_SUPPORTED`.

Assuming the default in the code is correct, shouldn't line 363 in your suggested code be `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED`? It would be confusing to change the default for `Py_mod_multiple_interpreters` when `Py_mod_gil` is explicitly set to its default value.

For line 364, we shouldn't make that assumption. I discussed this more in my reply to your `module.rst` comment, but I think `Py_MOD_GIL_NOT_USED` and `Py_MOD_MULTIPLE_INTERPRETERS_SUPPORTED` is the one combination that we should warn on or refuse to load. It's probably an oversight on the module author's part, and they most likely meant to give `Py_MOD_PER_INTERPRETER_GIL_SUPPORTED`, but the risk of subtle bugs from making that assumption seems too high.

Reply:
Yeah, I don't think there's anything to change here.